src/garage/torch/algos/maml.py

```diff
@@ -0,0 +1,475 @@
+"""Model-Agnostic Meta-Learning (MAML) algorithm implementation for RL."""
+import collections
+import copy
+
+from dowel import tabular
+import numpy as np
+import torch
+
+from garage.misc import tensor_utils
+from garage.sampler import OnPolicyVectorizedSampler
+from garage.tf.samplers import BatchSampler
+from garage.torch.algos import _Default, make_optimizer, VPG
+from garage.torch.optimizers import DiffSGD
+from garage.torch.utils import update_module_params
+
+
+class MAML:
+    """Model-Agnostic Meta-Learning (MAML).
+
+    Args:
+        env (garage.envs.GarageEnv): A gym environment.
+        policy (garage.torch.policies.Policy): Policy.
+        baseline (garage.np.baselines.Baseline): The baseline.
+        meta_optimizer (Union[torch.optim.Optimizer, tuple]):
+            Type of optimizer.
+            This can be an optimizer type such as `torch.optim.Adam` or a tuple
+            of type and dictionary, where dictionary contains arguments to
+            initialize the optimizer e.g. `(torch.optim.Adam, {'lr' = 1e-3})`.
+        meta_batch_size (int): Number of tasks sampled per batch.
+        inner_lr (double): Adaptation learning rate.
+        num_grad_updates (int): Number of adaptation gradient steps.
+        inner_algo (garage.torch.algos.VPG): The inner algorithm used for
+            computing loss.
+
+    """
+
+    def __init__(self,
+                 env,
+                 policy,
+                 baseline,
+                 meta_optimizer=torch.optim.Adam,
+                 meta_batch_size=40,
+                 inner_lr=0.1,
+                 num_grad_updates=1,
+                 inner_algo=None):
+        if inner_algo is None:
+            inner_algo = VPG(env.spec, policy, baseline)
+
+        if policy.vectorized:
+            self.sampler_cls = OnPolicyVectorizedSampler
+        else:
+            self.sampler_cls = BatchSampler
+
+        self.policy = policy
```

Comment:
please use `_private` for class members.

Reply:
Because runner has a reference to `algo.policy`, I will need to fix all algorithms. The same goes to `sampler_cls`.

Reply:
why not add a property?
```
@property
def policy(self):
    return self._policy
```

Reply:
I can do this if this is acceptable.